File: 581-shortest-unsorted-continuous-subarray/581-shortest-unsorted-continuous-subarray.cpp
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        return solve(nums);
    }
    
private:
    // Daily
    int solve(vector<int>& nums) {
        if(nums.size() == 1)
            return 0;
        
        stack<int> stack;
        
        int start = 1e6, end = -1;
        
        for(int i = 0; i < nums.size(); ++i) {
            while(!stack.empty() && nums[stack.top()] > nums[i]) {
                int idx = stack.top();
                stack.pop();
                start = min(start, idx);
            }
            stack.push(i);
        }
        
        while(!stack.empty())
            stack.pop();
        
        for(int i = nums.size() - 1; i >= 0; --i) {
            while(!stack.empty() && nums[stack.top()] < nums[i]) {
                int idx = stack.top();
                stack.pop();
                end = max(end, idx);
            }
            stack.push(i);
        }
        
        int ans = end - start + 1;
        if(ans < 0)
            return 0;
        return ans;
    }
};

File: 581-shortest-unsorted-continuous-subarray/581-shortest-unsorted-continuous-subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        start, end = len(nums), 0
        
        stack = deque()
        
        for i in range(len(nums)):
            while stack and nums[stack[-1]] > nums[i]:
                start = min(start, stack.pop())
            stack.append(i)
        
        stack = deque()
        
        for i in range(len(nums) - 1, -1, -1):
            while stack and nums[stack[-1]] < nums[i]:
                end = max(end, stack.pop())
            stack.append(i)
        
        ans = end - start + 1
        
        if ans < 0:
            return 0
        return ans
        